src/optimizer/query_to_operator_transformer.cpp

```diff
@@ -245,11 +249,62 @@ void QueryToOperatorTransformer::Visit(parser::InsertStatement *op) {
     op->select->Accept(this);
     insert_expr->PushChild(output_expr_);
     output_expr_ = insert_expr;
-  } else {
-    auto insert_expr = std::make_shared<OperatorExpression>(
-        LogicalInsert::make(target_table, &op->columns, &op->insert_values));
-    output_expr_ = insert_expr;
+    return;
+  }
+  // column_objects represents the columns for the current table as defined in
+  // its schema
+  auto column_objects = target_table->GetColumnObjects();
+  // INSERT INTO table_name VALUES (val1, val2, ...), (val_a, val_b, ...), ...
+  if (op->columns.empty()) {
+    for (uint32_t tuple_idx = 0; tuple_idx != op->insert_values.size(); ++tuple_idx) {
+      auto &values = op->insert_values[tuple_idx];
+      if (values.size() > column_objects.size()) {
+        throw CatalogException(
+            "ERROR:  INSERT has more expressions than target columns");
+      } else if (values.size() < column_objects.size()) {
+        for (oid_t i = values.size(); i != column_objects.size(); ++i) {  // check whether null values are allowed in the rest of the columns
+          if (column_objects[i]->IsNotNull()) {
+            // TODO: Add check for default value's existence for the current column
+            throw CatalogException(
+            "ERROR:  null value in column \"" + column_objects[i]->GetColumnName() + "\" violates not-null constraint");
+          }
+        }
+      }
+    }
+  }
+  // INSERT INTO table_name (col1, col2, ...) VALUES (val1, val2, ...), ...
+  else {
+    auto num_columns = op->columns.size();
+    // range based for-loop shows "call to implicitly-deleted copy
+    // constructor" error
+    for (uint32_t i = 0; i != op->insert_values.size();
+          ++i) {  // check size of each tuple
+      if (op->insert_values[i].size() > num_columns) {
+        throw CatalogException(
+            "ERROR:  INSERT has more expressions than target columns");
+      } else if (op->insert_values[i].size() < num_columns) {
+        throw CatalogException(
+            "ERROR:  INSERT has more target columns than expressions");
+      }
+    }
+
+    for (size_t idx = 0; idx != num_columns; ++idx) {
```

Comment:
Again, since you don't need `idx` aside from accessing elements, can you change this to a range-for loop with a `const` reference loop variable?